File: 03_liskov_substitution/typescript/good/index02-good.ts
class FlyingBird {
  fly (): void {
    console.log('I can fly')
  }
}

class SwimingBrid {
  swim (): void {
    console.log('I can swim')
  }
}

class Duck extends FlyingBird {
  quack (): void {
    console.log('I can quack')
  }
}
class Penguin extends SwimingBrid {

}

function makeFlyingBirdFly (bird: FlyingBird): void {
  bird.fly()
}
function makeSwimingBirdSwim (bird: SwimingBrid): void {
  bird.swim()
}

const duck = new Duck()
const penguin = new Penguin()

makeFlyingBirdFly(duck)
makeSwimingBirdSwim(penguin)

File: 03_liskov_substitution/dart/good/index02-good.dart
mixin FlyingBird {
  void fly() {
    print('I ${this.runtimeType} can fly');
  }
}

mixin SwimingBrid {
  void swim() {
    print('I ${this.runtimeType} can swim');
  }
}

class Duck with SwimingBrid, FlyingBird {
  void quack() {
    print('I ${this.runtimeType} can quack');
  }
}

class Penguin with SwimingBrid {}

void makeFlyingBirdFly(FlyingBird bird) {
  bird.fly();
}

void makeSwimingBirdSwim(SwimingBrid bird) {
  bird.swim();
}

void main(List<String> args) {
  var duck = Duck();
  var penguin = Penguin();

  makeFlyingBirdFly(duck);
  makeSwimingBirdSwim(duck);
  makeSwimingBirdSwim(penguin);
}
